1Password/shell-plugins
# feat: Add a Nix Flake with a devShell, NixOS module, and home-manager module to configure Shell Plugins

## Overview


The `flake.nix` added does two things:

1. Adds a `devShell` so Nix users can just run `nix develop` or use `direnv` to have a Go environment set up to build the plugins.
2. Add a NixOS module and a `home-manager` module (users will choose one or the other) to configure 1Password Shell Plugins via Nix; this only works if you're using Nix to manage your shell config, but if you're deep into Nix enough to be using Flakes, you probably are managing your shell config with Nix, too.

With this flake as an input to your own flake, you would then be able to configure 1Password Shell Plugins with the following:

```nix
programs._1password-shell-plugins = {
  enable = true;
  plugins = with pkgs; [ gh awscli2 cachix ];
};
```

## Type of change
 


## Related Issue(s)
  

* Resolves: #
* Relates: #

## How To Test


Create a `home-manager` Nix flake that manages your shell configuration:

`flake.nix`
```nix
{
  description = "My dotfiles managed with nix as a flake";
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    home-manager = {
      url = "github:nix-community/home-manager";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    # This should be the path to the repo checkout locally, with this branch checked out
    _1password-shell-plugins.url = "git+file:///$HOME/git/shell-plugins";
  };

  outputs = inputs@{ self, nixpkgs, home-manager, ... }: {
    homeConfigurations = {
      # note: this "my-config" name will be the value you use on the command line
      # to apply the configuration as well
      "my-config" = home-manager.lib.homeManagerConfiguration {
        extraSpecialArgs = { inherit inputs; };
        # or pkgs = nixpkgs.legacyPackages.x86_64-linux; if you're on linux
        pkgs = nixpkgs.legacyPackages.aarch64-darwin;
        modules = [
          inputs._1password-shell-plugins.hmModules.default
          ./home.nix
        ];
      };
    };
  };
}
```

`./home.nix`
```nix
{ inputs, config, pkgs, lib, ... }: {
  home = {
    username = "<your-os-username>";
    # or `/home/<your-linux-username>` if you're on Linux
    homeDirectory = "/Users/<your-os-username>";
    # this should be set up by home-manager automatically for you, leave it alone
    stateVersion = "22.11";
  };
  # allow 1Password CLI package; Nix doesn't allow any unfree software by default
  nixpkgs.config.allowUnfreePredicate = pkg:
    builtins.elem (lib.getName pkg) [
      "1password-cli"
    ];
  programs._1password-shell-plugins = {
    enable = true;
     # whatever shell plugins you want to use that have packages
     # available in nixpkgs
    plugins = with pkgs; [ gh ];
  };
  programs.fish = {
    enable = true;
    # the rest of your shell config goes here
  };
}
```

If you're on macOS, you will need to install `home-manager` manually as well: https://nix-community.github.io/home-manager/index.xhtml#sec-install-standalone

On NixOS, you can manage `home-manager` as a NixOS module instead.

For standalone `home-manage` installation, run `home-manager switch --flake ~/path/to/flake-directory/.#my-config` to apply the configuration. If you're on NixOS and using a flake to manage `home-manager` as a NixOS module, you can just run `sudo nixos-rebuild switch --flake ~/path/to/flake-directory/.#my-config`

## Changelog
  
Add a `flake.nix` that provides a `devShell` for contributors as well as a NixOS and `home-manager` module that can be used to configure 1Password Shell Plugins via Nix.
